File: src/main/java/photodb/service/rest/KeepAlive.java
/**
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package photodb.service.rest;

import photodb.cdi.DtoBuilder;
import photodb.data.dto.SessionDataDto;
import photodb.data.dto.UserInfo;
import photodb.service.bean.UserImpl;

import javax.inject.Inject;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpSession;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.Context;

@Path("/keep-alive")
@Produces("application/json")
public class KeepAlive {

    @Inject
    private DtoBuilder dtoBuilder;

    @Inject
    private UserImpl userService;

    @GET
    @Produces("application/json")
    @SuppressWarnings("PMD.EmptyCatchBlock")
    public SessionDataDto ping(@Context HttpServletRequest request) {
        final HttpSession session = request.getSession();
        final UserInfo info = this.dtoBuilder.build(this.userService.getUser());
        final String userName;
        final boolean logged;
        if (info == null) {
            userName = "guest";
            logged = false;
        } else {
            userName = info.getName();
            logged = true;
        }
        final SessionDataDto result = new SessionDataDto();
        result.setSessionId(session.getId());
        result.setUserName(userName);
        result.setLogged(logged);
        return result;
    }

}

File: src/main/webapp/app/js/keep-alive.js
/**
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 "use strict";
 */

define(['lib/underscore', 'app/js/log', 'lib/jquery'], function (underscore) {
    'use strict';

    var DELAY = 1000 * 60 * 4; // 4 minutes
    var timeoutKey = null;
    var callbacks = [];

    function scheduleNext(now) {
        if (timeoutKey !== null) {
            window.clearInterval(timeoutKey);
            window.console.log('keep-alive callback canceled.', timeoutKey);
            timeoutKey = null;
        }
        function timeoutCallback() {
            $.ajax({
                type: 'GET',
                'url': window.ux.ROOT_URL + 'rest/keep-alive',
                global: false,
                data: {},
                success: function (resp) {
                    var result = {
                        success: true,
                        data: resp.sessionDataDto
                    };
                    scheduleNext();
                    underscore.each(callbacks, function (cb) {
                        cb(result);
                    });
                },
                error: function () {
                    var result = {
                        success: false
                    };
                    underscore.each(callbacks, function (cb) {
                        cb(result);
                    });
                }
            });
        }

        if (now) {
            timeoutCallback();
        } else {
            timeoutKey = window.setTimeout(timeoutCallback, DELAY);
        }
        window.console.log('keep-alive callback created.', timeoutKey);
    }

    return {
        start: function () {
            $(window.document).bind("ajaxSend", function () {
                scheduleNext();
            });
            scheduleNext(true);
        },
        onPing: function (callback) {
            callbacks.push(callback);
        }
    };
});
